### ISSUE TITLE
fixed loss in training multi-class semantic task

### ISSUE THREAD

**Issue by Forest-Xie** [milesial/Pytorch-UNet#186]
Hi, I got some problems that I cannot find reasones, please give me some advice.
I modified the code to train with 21 classes, and I got normal training process. But for result, I got training loss about 0.2, and validation loss about 0.97, while I got mIoU lower than 0.0003. Given such a bad result, I modify the network output layer by adding a softmax layer, still I can start the training process, but I got training loss fixed at 0.5975 after a few epochs, I search some tips through Internet, but the advice is to add BN and use relu activation, which is already satisfied. I wonder whether it is ok to add BN and softmax for multi-class task? I cannot find the reason, maybe you can give me some suggestions? Thanks a lot, looking forward to your reply. Plus, I resize the image input to network for using bigger batch size, which I think will not effect the training a lot.

**Comment by Forest-Xie** [comment]
Thanks for you advice, the imbalanced classes problem do exist, I'll try what you said about class weights. Besides, does what you said "softmax is already used for multiclass", mean the code :
   with torch.no_grad():
        output = net(img)

        if net.n_classes > 1:
            output = F.softmax(output, dim=1)  # 如果使用argmax，并且没有反向传播的需求的话，这一步可省略
        else:
            output = torch.sigmoid(output)
in the file predice.py, for I just find softmax used here? If it is just used in predict.py, what about loss computation? Does not need? Or I ignore some detail? Looking forward to your reply. Thanks again.

**Comment by milesial** [comment]
It is also used for the train loss, but implicitly by the CrossEntropyLoss, which takes logits.
https://pytorch.org/docs/stable/nn.html#crossentropyloss

**Comment by Forest-Xie** [comment]
Thanks a lot, so sorry for I did not check the document carefully. I'll remove extra softmax layer, and I'm trying your advice for add class weight, hoping for a good result. Thanks for your explanation.
